File: 552-student-attendance-record-ii/student-attendance-record-ii.cpp
#define mod 1000000007
#define ll long long int

class Solution {
public:
    ll dp[100001][3][3];
    ll solve(int n,int a,int l){
        if(n==0)return 1;
        if(dp[n][a][l]!=-1){
            return dp[n][a][l];
        }
        ll ans=0;
        ans= solve(n-1,a,0) % mod;
        if(a<1) ans = (ans + solve(n-1,a+1,0)) % mod;
        if(l<2) ans = (ans + solve(n-1,a,l+1))%mod;
        return dp[n][a][l]= ans % mod;
    }

    int checkRecord(int n) {
        memset(dp,-1,sizeof(dp));
        return solve(n,0,0);
    }
};

File: 552-student-attendance-record-ii/student-attendance-record-ii.java
class Solution {
    public int checkRecord(int n) {
        int mod = 1_000_000_007;
        int[][][] dp = new int[100001][3][4];
        for(int a=0;a<=1;a++){
            for(int l=0;l<=2;l++){
                dp[0][a][l]=1;
            }
        }
        for(int day=1;day<=n;day++){
            for(int a=0;a<=1;a++){
                for(int l=0;l<=2;l++){
                    dp[day][a][l] = dp[day-1][a][0]%mod;
                    dp[day][a][l] = (dp[day][a][l] + dp[day-1][a+1][0])%mod;
                    dp[day][a][l] = (dp[day][a][l] + dp[day-1][a][l+1]) % mod;
                }
            }    
        }
        return dp[n][0][0];
    }
}